zetasql/fuzzing/component/runner.h

```diff
@@ -17,44 +17,24 @@
 #ifndef ZETASQL_FUZZING_RUNNER_H
 #define ZETASQL_FUZZING_RUNNER_H
 
-#include <memory>
 #include <functional>
-#include "zetasql/fuzzing/component/fuzz_targets/fuzz_target.h"
+#include <memory>
+
 #include "zetasql/fuzzing/component/arguments/argument.h"
+#include "zetasql/fuzzing/component/fuzz_targets/fuzz_target.h"
 
 namespace zetasql_fuzzer {
 
-template<typename InputType, typename TargetType, typename ... Functions>
+template <typename InputType, typename TargetType, typename... Functions>
 void Run(const InputType& input, Functions... functions) {
   TargetType target;
-  std::function<std::unique_ptr<Argument>(const InputType&)>
-      extractors[sizeof...(functions)] = {functions...};
-  for (auto& extractor : extractors) {
+  for (const std::function<std::unique_ptr<Argument>(const InputType&)>&
+           extractor : {functions...}) {
     extractor(input)->Accept(target);
```

Comment:
Should we change the semantics of extractor from `std::unique_ptr<Argument>(const InputType&)` to `void(const InputType&, FuzzTarget& target)`, i.e., extract_and_apply? In this way, we can eliminate the need to having a unique pointer here, since the Argument can be declared a local variable in each extract_and_apply function, and in each function we have something like 

```c++
void Run(const InputType& input, Functions... functions) {
  TargetType target;
  for (const std::function<void(const InputType&, FuzzTarget&)>&
           extract_and_apply : {functions...}) {
    extract_and_apply(input, target);
  }
  target.Execute();
}

void ExtractAndApply(const InputType& input, FuzzTarget& target) {
  SomeExtractor some_extractor;
  some_extractor.Extract(input);
  ConcreteArgType argument(some_extractor.Data());
  argument.Accept(target);
}
```

Reply:
I think either works, and the second does seem a bit neater (but less flexible, presumably).

I think it's probably up to you.  I'm not 100% sure we want void for the return type (can anything new go wrong in accept_and_apply instead of just accept?), but that's also something that can be addressed in some later iteration.

Reply:
Basically the difference is that the job of applying arguments to fuzz target is transferred from runner to the apply_and_extract function now.

Reply:
It also occurs to me that if we use this approach then we don't need the Visitor Pattern anymore, since the type information is preserved when we use local variables. Do you think there's value in keeping the visitor pattern @shumway @evmaus

Reply:
I'm also not sure if `ExtractAndApply` is the best example that fits single-responsibility principle

Reply:
I think you can drop the vistior pattern.

Reply:
You can leave the visitor pattern in for this CL and later remove as a refactoring if you want. Whichever works better for you.

Reply:
I can see that `ExtractAndApply` potentially gets misused however. Originally we're enforcing the function signatures to allow standard ways to getting the arguments from input to FuzzTarget, through #Visit interface. Having ExtractAndApply tightens the coupling, so that it's possible for future developers to create nonstandard interfaces unintentionally (for example, if I know FuzzTargetA takes ArgumentB, I can unintentionally creates FuzzTargetA#GetArgB which is not shared with other FuzzTarget subclass, since it wouldn't break the FuzzTarget interface), therefore decreases code reuse. Do you see this as a problem?

Reply:
This design choice is good enough for this CL.